### ISSUE TITLE
Incorrect ServiceType in predicate when using RegisterConditional with open generic registration

### ISSUE THREAD

**Issue by Fresa** [simpleinjector/SimpleInjector#333]
The following does not work as expected:

``` c#
public interface IConnection
{
    string ConnectionString { get; }
}

public class CarDatabaseConnection : IConnection
{
    public string ConnectionString { get; set; }
}

public class InMemoryDatabase<T> : IDatabase<T>
{
    private readonly IConnection _connection;

    public InMemoryDatabase(IConnection connection)
    {
        _connection = connection;
    }

    public List<T> Get(string type)
    {
        throw new NotImplementedException();
    }
}

public class Car
{
    
}

public interface IDatabase<T>
{
    List<T> Get(string type);
}

public class ServiceContainer
{
    public void Configure()
    {
        var container = new Container();

        container.RegisterConditional(
            typeof(IConnection),
            Lifestyle.Singleton.CreateRegistration(() => new CarDatabaseConnection(), container),
            c => c.Consumer != null && c.Consumer.ServiceType == typeof(IDatabase<Car>));

        container.Register(typeof(IDatabase<>), typeof(InMemoryDatabase<>), Lifestyle.Singleton);

        var database = container.GetInstance<IDatabase<Car>>();
    }
}
```

In the conditional predicate, both the ServiceType and the ImplementationType of the Consumer will be `InMemoryDatabase<Car>`. Shouldn't the ServiceType be `IDatabase<Car>`?

If I exchange `container.Register(typeof(IDatabase<>), typeof(InMemoryDatabase<>), Lifestyle.Singleton);` with `container.RegisterSingleton<IDatabase<Car>, InMemoryDatabase<Car>>();` the expected behavior is achived, i.e. the ServiceType in the predicate will be `IDatabase<Car>`.

EDIT: Tested in v3.3.0.

**Comment by dotnetjunkie** [comment]
Thanks for reporting this.

I can confirm this is a bug. This bug exists since v3.0 and is caused by [this (incorrect) line of code](https://github.com/simpleinjector/SimpleInjector/blob/v3.3.x/src/SimpleInjector/Internals/GenericRegistrationEntry.cs#L454). The fix however isn't as easy as it seems, because when this line is changed, [another test](https://github.com/simpleinjector/SimpleInjector/blob/v3.3.x/src/SimpleInjector.Core.Tests.Unit/RegisterConditionalTests.cs#L897) would fail.

I will look into this and try to push a patch release containing this fix.

**Comment by dotnetjunkie** [comment]
I just released v3.3.1 to NuGet that fixes this problem.
